src/librustc/ty/sty.rs

```diff
@@ -556,9 +556,39 @@ pub struct ProjectionTy<'tcx> {
     /// The trait reference `T as Trait<..>`.
     pub trait_ref: ty::TraitRef<'tcx>,
 
-    /// The name `N` of the associated type.
-    pub item_name: Name,
+    /// The DefId of the the associated type `N`.
+    ///
+    /// Note that this is not the DefId of the TraitRef containing this
+    /// associated type, which is in tcx.associated_item(item_def_id).container.
+    pub item_def_id: DefId,
 }
+
+impl<'a, 'tcx> ProjectionTy<'tcx> {
+    pub fn from_ref_and_name(
+        tcx: TyCtxt,
+        trait_ref: ty::TraitRef<'tcx>,
+        item_name: Name) -> ProjectionTy<'tcx> {
```

Comment:
Nit: formatting. I think that in this case, the `->` goes on next line, and I personally would put `{` indented with `pub` (so that body is clearly distinguished from the signature). But we don't have hard-and-fast rules here. You could also align with the open `(`.